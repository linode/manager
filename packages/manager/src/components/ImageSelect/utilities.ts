import { DateTime } from 'luxon';

import { MAX_MONTHS_EOL_FILTER } from 'src/constants';

import type { ImageSelectVariant } from './ImageSelect';
import type { Image, RegionSite } from '@linode/api-v4';
import type { DisableItemOption } from 'src/components/ListItemOption';

/**
 * Given a Image Select "variant", this PR returns an
 * API filter to help you fetch the correct images.
 */
export const getAPIFilterForImageSelect = (
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    return {
      '+order': 'desc',
      '+order_by': 'label',
      is_public: true,
    };
  }

  if (variant === 'private') {
    return { is_public: false };
  }

  return {};
};

/**
 * Unfortunately, we can't use API filtering for all of our filtering needs.
 *
 * This function exists to...
 * - filter out public Kubernetes images for the Images Select
 * - filter out images that are too far past their end-of-life date
 *
 * Please use API filtering (getAPIFilterForImageSelect) when possible!
 */
export const getFilteredImagesForImageSelect = (
  images: Image[] | undefined,
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    // For public images, we filter out LKE images and images that are > 6 months past their `eol`
    return images?.filter(
      (image) => !image.id.includes('kube') && !isImageTooFarPastEOL(image)
    );
  }

  return images;
};

/**
 * Returns whether or not an image is too far past its end-of-life based on MAX_MONTHS_EOL_FILTER
 *
 * This function is intended to be used to filter out end-of-life images based on
 * an internal policy. See `M3-5753` for context.
 *
 * @param image an image from the API
 * @returns true if the given image is more than 6 months past its end-of-life
 */
export const isImageTooFarPastEOL = (image: Image) => {
  if (image.eol === null) {
    return false;
  }

  const imageEOL = DateTime.fromISO(image.eol);
  const now = DateTime.now();

  const differenceInMonths = now.diff(imageEOL, 'months').months;

  return differenceInMonths > MAX_MONTHS_EOL_FILTER;
};

/**
 * Returns whether or not an image is deprecated
 *
 * Unfortunately, as per `M3-5753`, we can't just simply check `deprecated` on the `Image`,
 * we must also consider the EOL date.
 */
export const isImageDeprecated = (image: Image) => {
  if (image.eol === null) {
    // If the image does not have an EOL, just use the `deprecated` field returned by the API
    return image.deprecated;
  }

  const isImageEOL = DateTime.fromISO(image.eol) < DateTime.now();

  return image.deprecated || isImageEOL;
};

interface DisabledImageOptions {
  images: Image[];
  site_type?: RegionSite;
}

/**
 * Returns images that should be disabled on the Linode Create flow.
 *
 * @returns key/value pairs for disabled images. the key is the image id and the value is why the image is disabled
 */
export const getDisabledImages = (options: DisabledImageOptions) => {
  const { images, site_type } = options;

  // Disable images that do not support distributed sites if the selected region is distributed
  if (site_type === 'distributed') {
    const disabledImages: Record<string, DisableItemOption> = {};
    for (const image of images) {
      if (!image.capabilities.includes('distributed-sites')) {
        disabledImages[image.id] = {
          reason:
            'The selected image cannot be deployed to a distributed region.',
        };
      }
    }
    return disabledImages;
  }

  return {};
};
