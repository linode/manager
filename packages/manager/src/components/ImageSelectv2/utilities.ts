import { DateTime } from 'luxon';

import { MAX_MONTHS_EOL_FILTER } from 'src/constants';

import type { ImageSelectVariant } from './ImageSelectv2';
import type { Image } from '@linode/api-v4';

/**
 * Given a Image Select "variant", this PR returns an
 * API filter to help you fetch the correct images.
 */
export const getAPIFilterForImageSelect = (
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    return { '+order': 'desc', '+order_by': 'label', is_public: true };
  }

  if (variant === 'private') {
    return { is_public: false };
  }

  return {};
};

/**
 * Unfortunately, we can't use API filtering for all of our filtering needs.
 *
 * This function exists to...
 * - filter out public Kubernetes images for the Images Select
 * - filter out images that are too far past their end-of-life date
 *
 * Please use API filtering (getAPIFilterForImageSelect) when possible!
 */
export const getFilteredImagesForImageSelect = (
  images: Image[] | undefined,
  variant: ImageSelectVariant | undefined
) => {
  if (variant === 'public') {
    // For public images, we filter out LKE images and images that are > 6 months past their `eol`
    return images?.filter(
      (image) => !image.id.includes('kube') && !isImageTooFarPastEOL(image)
    );
  }

  return images;
};

/**
 * Returns whether or not an image is too far past its end-of-life based on MAX_MONTHS_EOL_FILTER
 *
 * This function is intended to be used to filter out end-of-life images based on
 * an internal policy. See `M3-5753` for context.
 *
 * @param image an image from the API
 * @returns true if the given image is more than 6 months past its end-of-life
 */
export const isImageTooFarPastEOL = (image: Image) => {
  if (image.eol === null) {
    return false;
  }

  const imageEOL = DateTime.fromISO(image.eol);
  const now = DateTime.now();

  const differenceInMonths = now.diff(imageEOL, 'months').months;

  return differenceInMonths > MAX_MONTHS_EOL_FILTER;
};

/**
 * Returns whether or not an image is deprecated
 *
 * Unfortunately, as per `M3-5753`, we can't just simply check `deprecated` on the `Image`,
 * we must also consider the EOL date.
 */
export const isImageDeprecated = (image: Image) => {
  if (image.eol === null) {
    // If the image does not have an EOL, just use the `deprecated` field returned by the API
    return image.deprecated;
  }

  const isImageEOL = DateTime.fromISO(image.eol) < DateTime.now();

  return image.deprecated || isImageEOL;
};
