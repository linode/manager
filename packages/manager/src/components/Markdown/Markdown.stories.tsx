import React from 'react';

import { Markdown } from './Markdown';

import type { Meta, StoryObj } from '@storybook/react';

const markdown = `
# h1 Heading
## h2 Heading
### h3 Heading
#### h4 Heading
##### h5 Heading
###### h6 Heading

**This is bold text**

*This is italic text*

1. Make my changes
    1. Fix bug
    2. Improve formatting
        - Make the headings bigger
2. Push my commits to GitHub
3. Open a pull request
    * Describe my changes
    * Mention all the members of my team
        * Ask for feedback

This in inline \`testing\` code

\`\`\`js
const fetchData = async (): Promise => {
  // ...
};
\`\`\`

| Month    | Savings |
| -------- | ------- |
| January  | $250    |
| February | $80     |
| March    | $420    |
`;

const meta: Meta<typeof Markdown> = {
  args: {
    textOrMarkdown: markdown,
  },
  component: Markdown,
  title: 'Components/Markdown',
};

export default meta;

type Story = StoryObj<typeof Markdown>;

export const Default: Story = {
  render: (args) => <Markdown {...args} />,
};
