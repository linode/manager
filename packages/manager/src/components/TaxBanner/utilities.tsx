import * as React from 'react';
import { Link } from 'react-router-dom';

export interface LinkType {
  link: string;
  text_to_replace: string;
  type: 'internal' | 'external';
}

export const generateHTMLFromString = (
  text: string,
  links: Record<string, LinkType>
) => {
  return Object.keys(links).reduce((acc, eachKey) => {
    const { text_to_replace, type, link } = links[eachKey];

    if (acc.length > 0) {
      // nth iteration after the first
      return acc.map(eachValue => {
        if (typeof eachValue === 'string') {
          return wrapStringInLink(eachValue, text_to_replace, type, link);
        } else {
          /** is a react component so no need to wrap anything in a link */
          return eachValue;
        }
      });
    } else {
      return wrapStringInLink(text, text_to_replace, type, link);
    }
  }, []);
};

export const wrapStringInLink = (
  fullString: string,
  textToReplace: string,
  type: 'internal' | 'external',
  link: string
) => {
  /**
   * split the sentence up by phrases, including the matched text in the array
   *
   * For example:
   *
   * While "I am" being the text to match, "hello world I am Marty"
   *
   * becomes
   *
   * ["hello world", "I am", "Marty"]
   *
   * See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split#Splitting_with_a_RegExp_to_include_parts_of_the_separator_in_the_result
   * for information about splitting but keeping the matched value in the array
   *
   */
  const sentenceAsArray = fullString.split(
    new RegExp(`(${textToReplace})`, 'gmi')
  );

  /**
   * iterate over the array and if we have a matched word, wrap
   * in an anchor tag or react-router link depending on the link type
   */
  const arrayWithLink = sentenceAsArray.map(eachWord => {
    if (eachWord.match(new RegExp(textToReplace, 'gmi'))) {
      return type === 'external' ? (
        <a
          key={eachWord}
          href={link}
          target="_blank"
          aria-describedby="external-site"
          rel="noopener noreferrer"
        >
          {eachWord}
        </a>
      ) : (
        <Link key={eachWord} to={link}>
          {eachWord}
        </Link>
      );
    }
    return eachWord;
  });

  /**
   * so now we have an array that looks like
   * [
   *  "hello world",
   *  "<a>I am</a>"
   *  "Marty"
   * ]
   */
  return arrayWithLink;
};
