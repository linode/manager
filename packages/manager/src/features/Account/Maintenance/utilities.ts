export const PENDING_MAINTENANCE_FILTER = Object.freeze({
  status: { '+or': ['pending', 'started'] },
});
