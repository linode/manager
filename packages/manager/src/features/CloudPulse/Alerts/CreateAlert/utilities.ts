import { omitProps } from '@linode/ui';

import type { CreateAlertDefinitionForm, MetricCriteriaForm } from './types';
import type {
  CreateAlertDefinitionPayload,
  MetricCriteria,
} from '@linode/api-v4';

// filtering out the form properties which are not part of the payload
export const filterFormValues = (
  formValues: CreateAlertDefinitionForm
): CreateAlertDefinitionPayload => {
  const values = omitProps(formValues, [
    'serviceType',
    'region',
    'engineType',
    'severity',
    'rule_criteria',
  ]);
  // severity has a need for null in the form for edge-cases, so null-checking and returning it as an appropriate type
  const severity = formValues.severity ?? 1;
  const entityIds = formValues.entity_ids;
  const rules = formValues.rule_criteria.rules;
  return {
    ...values,
    entity_ids: entityIds,
    rule_criteria: { rules: filterMetricCriteriaFormValues(rules) },
    severity,
  };
};

export const filterMetricCriteriaFormValues = (
  formValues: MetricCriteriaForm[]
): MetricCriteria[] => {
  return formValues.map((rule) => {
    const values = omitProps(rule, ['aggregation_type', 'operator', 'metric']);
    return {
      ...values,
      aggregation_type: rule.aggregation_type ?? 'avg',
      dimension_filters: rule.dimension_filters,
      metric: rule.metric ?? '',
      operator: rule.operator ?? 'eq',
    };
  });
};

export const convertToSeconds = (secondsList: string[]) => {
  return secondsList.map((second) => {
    const unit = second.slice(-1)[0];
    const number = parseInt(second.slice(0, -1), 10);
    switch (unit) {
      case 's':
        return number;
      case 'm':
        return number * 60;
      case 'h':
        return number * 3600;
      default:
        return number * 0;
    }
  });
};
