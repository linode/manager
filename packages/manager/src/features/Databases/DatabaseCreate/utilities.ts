export const determineReplicationType = (
  clusterSize: number,
  engine: string
) => {
  if (Boolean(engine.match(/mongo/))) {
    return undefined;
  }

  // If engine is a MySQL or Postgres one and it's a standalone DB instance
  if (clusterSize === 1) {
    return 'none';
  }

  // MySQL engine & cluster = semi_synch. PostgreSQL engine & cluster = asynch.
  if (Boolean(engine.match(/mysql/))) {
    return 'semi_synch';
  } else {
    return 'asynch';
  }
};

export const determineReplicationCommitType = (engine: string) => {
  // 'local' is the default.
  if (Boolean(engine.match(/postgres/))) {
    return 'local';
  }

  return undefined;
};

export const determineStorageEngine = (engine: string) => {
  // 'wiredtiger' is the default.
  if (Boolean(engine.match(/mongo/))) {
    return 'wiredtiger';
  }

  return undefined;
};

export const determineCompressionType = (engine: string) => {
  // 'none' is the default.
  if (Boolean(engine.match(/mongo/))) {
    return 'none';
  }

  return undefined;
};
