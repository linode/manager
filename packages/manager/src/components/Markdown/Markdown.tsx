import { Typography } from '@linode/ui';
import { useTheme } from '@mui/material';
import MarkdownIt from 'markdown-it';
import React from 'react';

import { sanitizeHTML } from 'src/utilities/sanitizeHTML';
import { getHighlighterTheme, shiki } from 'src/utilities/syntax-highlighter';

import type { SanitizeOptions } from 'src/utilities/sanitizeHTML';

export interface HighlightedMarkdownProps {
  className?: string;
  sanitizeOptions?: SanitizeOptions;
  textOrMarkdown: string;
}

/**
 * Renders markdown as HTML using https://github.com/markdown-it/markdown-it
 * - Used mostly for rendering support ticket messages
 * - Will perform syntax highlighting on any fenced code blocks
 */
export const Markdown = (props: HighlightedMarkdownProps) => {
  const { className, sanitizeOptions, textOrMarkdown } = props;

  const theme = useTheme();

  // This is "unsafe" because the output may contain HTML. To be deemed "safe" it
  // needs to be sanitized before rendered as HTML.
  const unsafeMarkdownIt = new MarkdownIt({
    /**
     * We pass a custom highlight function to perform syntax highlighting on code blocks
     */
    highlight(str, lang) {
      try {
        return shiki.codeToHtml(str, {
          lang,
          theme: getHighlighterTheme(theme.palette.mode),
        });
      } catch (error) {
        return shiki.codeToHtml(str, {
          lang: 'js',
          theme: getHighlighterTheme(theme.palette.mode),
        });
      }
    },
    html: true,
    linkify: true,
  });

  const unsafeParsedMarkdown = unsafeMarkdownIt.render(textOrMarkdown);

  const sanitizedHtml = sanitizeHTML({
    allowMoreAttrs: ['class', 'style'],
    allowMoreTags: ['code', 'span', 'pre'],
    sanitizeOptions,
    sanitizingTier: 'flexible',
    text: unsafeParsedMarkdown,
  });

  return (
    <Typography
      className={className}
      dangerouslySetInnerHTML={{ __html: sanitizedHtml }}
      sx={{ pre: { borderRadius: 1, overflowX: 'auto', p: 1 } }}
    />
  );
};
