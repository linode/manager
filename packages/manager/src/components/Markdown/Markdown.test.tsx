import * as React from 'react';

import { renderWithTheme } from 'src/utilities/testHelpers';

import { Markdown } from './Markdown';

const sampleMarkdown =
  '# Some markdown \n ```javascript\n const x = function() { return true; }\n```';

describe('Markdown component', () => {
  it('should highlight text consistently', () => {
    const { asFragment } = renderWithTheme(
      <Markdown textOrMarkdown={sampleMarkdown} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
