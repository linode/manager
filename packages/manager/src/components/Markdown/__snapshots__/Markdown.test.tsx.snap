// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Markdown component > should highlight text consistently 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1 css-gqbs1y-MuiTypography-root"
  />
  <h1>
    Some markdown
  </h1>
  

  <pre
    class="shiki one-light"
    style="background-color:#FAFAFA;color:#383A42"
  >
    <code>
      <span
        class="line"
      >
        <span
          style="color:#A626A4"
        >
          const
        </span>
        <span
          style="color:#4078F2"
        >
           x
        </span>
        <span
          style="color:#0184BC"
        >
           =
        </span>
        <span
          style="color:#A626A4"
        >
           function
        </span>
        <span
          style="color:#383A42"
        >
          () { 
        </span>
        <span
          style="color:#A626A4"
        >
          return
        </span>
        <span
          style="color:#986801"
        >
           true
        </span>
        <span
          style="color:#383A42"
        >
          ; }
        </span>
      </span>
      

      <span
        class="line"
      />
    </code>
  </pre>
  <p />
</DocumentFragment>
`;
